pkgs/by-name/ma/material-maker/package.nix

```diff
@@ -0,0 +1,157 @@
+  libXinerama,
+  libXrandr,
+  libXrender,
+  alsa-lib,
+  libpulseaudio,
+  udev,
+}:
+let
+  pname = "material-maker";
+  version = "1.3";
+  versionUnderscore = "${lib.replaceStrings [ "." ] [ "_" ] version}";
+
+  sources = {
+    linux = fetchzip {
+      url = "https://github.com/RodZill4/material-maker/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
+      hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
+    };
+    darwin = fetchurl {
+      url = "https://github.com/RodZill4/material-maker/releases/download/${version}/material_maker_${versionUnderscore}.dmg";
+      hash = "sha256-D4jPQEOKws2JFczh0K55mR08jIBMU6jOFOm6TyFxnt8=";
+    };
+  };
+in
+stdenvNoCC.mkDerivation (finalAttrs: {
+  inherit pname version;
+
+  src =
+    if stdenvNoCC.isLinux && stdenvNoCC.isx86_64 then
+      sources.linux
+    else if stdenvNoCC.isDarwin then
+      sources.darwin
+    else
+      throw "unsupported platform";
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    copyDesktopItems
+  ] ++ lib.optionals stdenvNoCC.isDarwin [ undmg ];
+
+  sourceRoot = ".";
+
+  buildInputs = [
+    libGL
+    libGLU
+    libX11
+    libXcursor
+    libXext
+    libXfixes
+    libXi
+    libXinerama
+    libXrandr
+    libXrender
+  ];
+
+  desktopItems = makeDesktopItem {
+    name = "material-maker";
+    exec = "material-maker";
+    icon = "material-maker";
+    genericName = "Procedural texture generation and 3D model painting tool";
+    desktopName = "Material Maker";
+    comment = finalAttrs.meta.description;
+    terminal = false;
+    categories = [
+      "Graphics"
+      "3DGraphics"
+    ];
+    keywords = [
+      "material"
+      "textures"
+      "paint"
+      "3d"
+    ];
+  };
+  
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenvNoCC.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
+      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck
+
+      for dir in library nodes environments meshes export; do
+        cp -R $src/$dir $out/opt
+      done
+
+      cp -R $src/examples $examples
+
+      mkdir -p $out/share/
+      cp -R $src/doc $out/share
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/opt/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString (builtins.elem "doc" finalAttrs.meta.outputsToInstall && stdenvNoCC.isLinux) ''
+      ln -s $doc/share/doc $out/opt/doc
+    ''
+    +
+      lib.optionalString (builtins.elem "examples" finalAttrs.meta.outputsToInstall && stdenvNoCC.isLinux)
+        ''
+          ln -s $examples $out/opt/examples
+        ''
+    + lib.optionalString stdenvNoCC.isDarwin ''
+      mkdir -p $out/Applications
+      find ./ -type d -name "*.app" -exec cp -r {} /destination/directory $out/Applications \;
+    ''
+    + ''
+      runHook postInstall
+    '';
```

Comment:
In my opinion, looping over `getAllOutputNames` is more readable.

Also, since the command that installs the icon also creates the `$outshare` directory, we can move `cp -R $src/doc $out/share` below that so we don't have to `mkdir -p $out/share/`

```suggestion
  installPhase =
    ''
      runHook preInstall
    ''
    + lib.optionalString stdenvNoCC.isLinux ''
      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck

      for dir in library nodes environments meshes export; do
        cp -R $src/$dir $out/opt
      done

      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

      mkdir -p $out/bin
      ln -s $out/opt/material-maker $out/bin/material-maker

      cp -R $src/doc $out/share
      cp -R $src/examples $examples

      for output in $(getAllOutputNames); do
        if [ "$output" == "doc" ]; then
          ln -s $doc/share/doc $out/opt/doc
        elif [ "$output" == "examples" ]; then
          ln -s $examples $out/opt/examples
        fi
      done
    ''
    + lib.optionalString stdenvNoCC.isDarwin ''
      mkdir -p $out/Applications
      find ./ -type d -name "*.app" -exec cp -r {} /destination/directory $out/Applications \;
    ''
    + ''
      runHook postInstall
    '';
```

Reply:
I added it but changed it so it queries outputsToInstall instead, i think this is wat it needs to follow, correct me if im wrong

Reply:
Because if the user wants to only install the out and not doc and examples they will override the meta.outputsToInstall because if you remove outputs it doesnt build anymore